exonum/src/runtime/dispatcher/mod.rs

```diff
@@ -409,23 +593,147 @@ impl Dispatcher {
+                schema.add_local_migration_result(instance_name, local_result);
+            }
+        }
+
+        // Rollback migrations.
+        let migration_rollbacks =
+            modified_instances
+                .iter()
+                .filter_map(|(state, prev_status)| match (&state.status, prev_status) {
+                    (Some(InstanceStatus::Stopped), Some(InstanceStatus::Migrating { .. })) => {
+                        Some(&state.spec.name)
+                    }
+                    _ => None,
+                });
+        for namespace in migration_rollbacks {
+            // Remove the thread corresponding to the migration (if any). This will abort
+            // the migration script since its `AbortHandle` is dropped.
+            self.migrations.threads.remove(namespace);
+            rollback_migration(&mut fork, namespace);
+        }
+
+        // Check if any migrations have finished. Record migration results in the DB.
+        let results = self.migrations.take_completed();
+        let mut schema = Schema::new(&fork);
+        for (instance_name, result) in results {
+            schema.add_local_migration_result(&instance_name, result);
+        }
+
         let patch = fork.into_patch();
 
         // Block futures with pending deployments.
         for (artifact, deploy_spec) in pending_artifacts {
             self.block_until_deployed(artifact, deploy_spec);
         }
+
         // Notify runtime about changes in service instances.
-        for (spec, status) in modified_instances {
-            self.update_service_status(&patch, &spec, status);
+        for (state, _) in modified_instances {
+            let status = state
+                .status
+                .expect("BUG: Service status cannot be changed to `None`");
+            self.update_service_status(&patch, &state.spec, status.clone());
+            if let Some(target) = status.ongoing_migration_target() {
+                self.start_migration_script(target, state.spec);
+            }
         }
+
         patch
     }
 
+    fn get_migration_script(
+        &self,
+        new_artifact: &ArtifactId,
+        old_service: &InstanceSpec,
+    ) -> Result<Option<MigrationScript>, ExecutionError> {
+        let runtime = self
+            .runtime_by_id(new_artifact.runtime_id)
+            .ok_or(Error::IncorrectRuntime)?;
+        runtime
+            .migrate(new_artifact, &old_service)
+            .map_err(From::from)
+    }
+
+    fn start_migration_script(&mut self, new_artifact: &ArtifactId, old_service: InstanceSpec) {
+        let script = self
+            .get_migration_script(new_artifact, &old_service)
+            .unwrap_or_else(|err| {
+                panic!(
+                    "BUG: Cannot obtain migration script for migrating {:?} to new artifact {:?}, {}",
+                    old_service, new_artifact, err
+                );
+            })
+            .unwrap_or_else(|| {
```

Comment:
Two chained `unwrap_or_else` look a bit odd. Maybe store intermediate result, so it will be easier to understand what've obtained after first `unwrap` and why we should `unwrap` again?